systemd-sleep: prefer resume device or file

zachsmith opened the pull request:
This change checks each swap partition or file reported in /proc/swaps
to see if it matches the values configured with resume= and
resume_offset= kernel parameters. If a match is found, the matching swap
entry is used as the hibernation location regardless of swap priority.

Fixes https://github.com/systemd/systemd/issues/13681

zachsmith commented on src/shared/sleep-config.c:
```diff
@@ -229,6 +245,50 @@ int find_hibernate_location(char **device, char **type, uint64_t *size, uint64_t
+                                        log_debug("Unable to detect offset on Btrfs: %s", swap->device);
+
+                                // If the resume device is the same as the swap file device, check the offset.
+                                if (streq(resume, swap_device)) {
```
This should be an `else if` to the previous `if` clause to ensure this doesn't run if on `Btrfs`.

zachsmith commented:
@evverx You helped me in the past with a CI issue so I'm tagging you here - please let me know if I should look elsewhere for help on the following question:

I looks like the `bionic-arm64` test failed for what I think is a CI timeout. Is it best for me to push again to trigger a re-run of everything or is there a way I can re-run just the failed test?

evverx commented:
As far as I can tell, `TEST-40-EXEC-COMMAND-EX` has been failing on Ubuntu CI for a while. The first failure with `Main process exited, code=exited, status=2/INVALIDARGUMENT` was kind of spotted in https://github.com/systemd/systemd/pull/13579#issuecomment-532721171 and I don't think we've figured it out since then. I wouldn't rerun the CI because I think it has nothing to do with this PR. It would probably make sense to skip it altogether. @ddstreet could you take a look?

evverx commented:
@anitazha before we start turning the test off, since it was added in https://github.com/systemd/systemd/pull/13369 and was merged even though it failed there with the same error message, I'm wondering if you could take a look at what is going on. Thank you!

anitazha commented:
> @anitazha before we start turning the test off, since it was added in #13369 and was merged even though it failed there with the same error message, I'm wondering if you could take a look at what is going on. Thank you!

Sure I will look into it today

ddstreet commented:
> . It would probably make sense to skip it altogether. @ddstreet could you take a look?

It failed for a PR of mine also; I haven't had a change to look into the specifics of why it's failing but it does seem to reliably fail on the arm64 test.

I'll get back to #13568 tomorrow and get that, and the associated Debian MR, out of WIP status so we can merge them, it seems like everyone is in agreement that's a good direction to go, and that would allow (temporarily) blacklisting specific tests for specific archs, like this one.

anitazha commented:
I managed to repro the test failure locally (`testsuite.service: Main process exited, code=exited, status=2/INVALIDARGUMENT`) by inserting `sleep 5` into `testsuite.sh` so I guess it's timing out (don't know why it exits with 2 though). Is there a way to increase the timeout or is disabling the right way to go?

evverx commented:
If the test is timing out, the timeout should (and most likely can) be increased somewhere but I have to admit I have no idea what kind of timeout can kick in there. Ideally we shouldn't sweep issues under the rug by turning tests off but if they fail regularly and nobody looks into it, it usually indicates nobody is particularly interested in getting them fixed and they can safely be turned off so as not to annoy contributors at least. As far as I can see, @mrc0mmand opened https://github.com/systemd/systemd/pull/13692. Let's see how it goes. I'd also wait for https://github.com/systemd/systemd/pull/13568 to land to turn the test off properly.

keszybz commented on src/shared/sleep-config.c:
```diff
@@ -200,6 +204,18 @@ int find_hibernate_location(char **device, char **type, uint64_t *size, uint64_t
+                else
+                        return log_debug_errno(r, "Error reading from /sys/power/resume_offset: %m");
+        }
+
```
This copies a few lines of code from configure_hibernation(). Please don't copy copy, but instead turn it into a shared function — see the main comment above.

keszybz commented on src/shared/sleep-config.c:
```diff
@@ -229,6 +245,51 @@ int find_hibernate_location(char **device, char **type, uint64_t *size, uint64_t
                                 continue;
                         }
 
+                        /* Check if swap matches the resume= and resume_offset= values if configured. */
```
Please let this be a separate function. It'll be much clearer and error handling will be easier.

keszybz commented on src/shared/sleep-config.c:
```diff
@@ -229,6 +245,51 @@ int find_hibernate_location(char **device, char **type, uint64_t *size, uint64_t
 
+                        /* Check if swap matches the resume= and resume_offset= values if configured. */
+                        if (!streq(resume_offset, "0") && !streq(resume, "0:0")) {
+                                _cleanup_close_ int fd;
```
Everything declared with `_cleanup_` must be initialized (except for some cases in test code, but that is not relevant here). I.e. you need `= -1;` here.

keszybz commented on src/shared/sleep-config.c:
```diff
@@ -229,6 +245,51 @@ int find_hibernate_location(char **device, char **type, uint64_t *size, uint64_t
+                                if (r < 0)
+                                        return log_error_errno(errno, "Failed to stat %s: %m", swap->device);
+
+                                if (asprintf(&swap_device, "%u:%u", major(sb.st_dev), minor(sb.st_dev)) < 0)
```
No need to use `asprintf`. Instead do
```C
char swap_device[DECIMAL_STR_MAX(dev_t)*2 + 1];
...
xsprintf(swap_device, "%u:%u", major(sb.st_dev), minor(sb.st_dev));
```

keszybz commented on src/shared/sleep-config.c:
```diff
@@ -229,6 +245,51 @@ int find_hibernate_location(char **device, char **type, uint64_t *size, uint64_t
+                                btrfs = btrfs_is_filesystem(fd);
+                                if (btrfs < 0)
+                                        return log_error_errno(r, "Error checking %s for Btrfs filesystem: %m", swap->device);
+                                else if (btrfs)
```
`btrfs > 0` is preferred when `btrfs` is an integer.

keszybz commented on src/shared/sleep-config.c:
```diff
@@ -229,6 +245,51 @@ int find_hibernate_location(char **device, char **type, uint64_t *size, uint64_t
+                                                log_oom();
+
+                                        /* If the swapfile offset matches the resume_offset, use it. */
+                                        if (streq(resume_offset, offset_str)) {
```
Please don't do comparisons using strings. Convert "resume=" into a integer and use that for comparisons.

keszybz commented on src/shared/sleep-config.c:
```diff
@@ -237,6 +298,24 @@ int find_hibernate_location(char **device, char **type, uint64_t *size, uint64_t
+
+                        /* Check if swap matches the resume= value if configured. */
+                        if (!streq(resume, "0:0")) {
+                                _cleanup_free_ char *swap_partition;
```
Same here. This will crash if the function returns before swap_partition has been intiialized. But you don't need this, use a static buffer instead and xsprintf.

zachsmith commented:
> Current state is that `find_hibernate_location` is called from `enough_swap_for_hibernation()` and ` write_hibernate_location_info()`. But your patch makes the code even more smelly by duplicating parts of the discovery code. Please unentangle this first:

Thanks for the detailed review, @keszybz and the for the encouragement to refactor further. As is probably evident, I'm learning as I go and don't always know how far to push the changes. :smile: 

> let's have one function which does the discovery and autoconfiguration of hibernation location. `write_hibernation_location_info()` should do no discovery on its own, just be passed the stuff to write as arguments. 

This makes good sense to me. I've cleaned up `write_hibernation_location_info` to receive all the information via parameter so was able to clean it up significantly.

> My suggestion would be to rename `find_hibernate_location()` to `static guess_hibernate_location()` and then add new `find_hibernate_location()` that would do the discovery that `write_hibernate_location_info()` currently does, and if that doesn't work, call `guess_hibernate_location()`. 

I tried this approach initially, but because `find_hibernate_location` is checking for the `/proc/swaps` entry that matches the `resume=` and `resume_offset` as it parses, it wasn't as straightforward to decouple. Detecting a `/proc/swaps` entry that matches `resume` and `resume_offset` is necessary to insure that `enough_swap_for_hibernation` can inspect the correct swap - and is the primary point of this patch. The approach I took was to factor some of the logic out of `find_hibernate_location` into separate functions to improve readability (as you suggested) and introduced a `HibernationLocation` struct that is constructed by `find_hibernation_location` that contains the all of the information needed by `write_hibernation_location_info` and `enough_swap_for_hibernation`.

> The final goal is to have just one place in the code which reads `/sys/power/resume` and `/sys/power/resume_offset`.

I was able to accomplish this and was able to remove a bit of code along the way. This should be DRY'ed up a bit better now.

Let me know what you think.

keszybz commented on src/shared/sleep-config.c:
```diff
@@ -237,45 +384,58 @@ int find_hibernate_location(char **device, char **type, uint64_t *size, uint64_t
+                                return r;
+
+                        hibernate_location = hibernate_location_free(hibernate_location);
+                        r = hibernate_location_new(TAKE_PTR(swap), TAKE_PTR(swap_device_id), swap_offset, &hibernate_location);
```
If hibernate_location_new fails, this will leak both pointers. I think it'd be easier to just allocate the structure here.

keszybz commented on src/shared/sleep-config.c:
```diff
@@ -237,45 +384,58 @@ int find_hibernate_location(char **device, char **type, uint64_t *size, uint64_t
+                                return r;
+
+                        /* if the swap is the resume device, stop looping swaps */
+                        if (location_is_resume_device(hibernate_location, sys_resume, sys_offset) > 0)
```
No "> 0" after this returns a `bool`.

keszybz commented on src/shared/sleep-config.c:
```diff
@@ -184,12 +177,145 @@ static SwapEntry* swap_entry_free(SwapEntry *se) {
+        if (btrfs < 0)
+                return log_error_errno(r, "Error checking %s for Btrfs filesystem: %m", swap->device);
+        else if (btrfs > 0) {
+                log_debug("Detection of swap file offset on Btrfs is not supported: %s; skipping", swap->device);
```
So here we have `return 0` and nothing is set. The general rule is: non-negative return == success == all outputs are set. Please either return negative here or set the output variable.

keszybz commented on src/shared/sleep-config.c:
```diff
@@ -184,12 +177,145 @@ static SwapEntry* swap_entry_free(SwapEntry *se) {
+
+        if (!streq(swap->type, "file")) {
+                log_warning("Attempted to calculate offset on swap entry that is not of type 'file': '%s'", swap->type);
+                *ret_offset = 0;
```
This is a static function, so `log_warning` is not useful. Just do an `assert`.

keszybz commented on src/shared/sleep-config.c:
```diff
@@ -184,12 +177,145 @@ static SwapEntry* swap_entry_free(SwapEntry *se) {
+ * Returns:
+ *  1 - HibernateLocation matches values found in /sys/power/resume & /sys/power/resume_offset
+ *  0 - HibernateLocation is highest priority swap with most remaining space; no valid values exist in /sys/power/resume & /sys/power/resume_offset
+ *  negative value in the case of error
```
Nice.

zachsmith commented:
@keszybz I think I’ve gotten through all the feedback - thanks again for all the tips and help. Let me know if there is anything else that needs to be addressed. :+1:
